h2o-core/src/main/java/water/H2O.java

```diff
@@ -793,8 +794,61 @@ private static void validateArguments() {
   /**
    * Register embedded H2O configuration object with H2O instance.
    */
-  public static void setEmbeddedH2OConfig(AbstractEmbeddedH2OConfig c) { embeddedH2OConfig = c; }
-  public static AbstractEmbeddedH2OConfig getEmbeddedH2OConfig() { return embeddedH2OConfig; }
+  public static void setEmbeddedH2OConfig(AbstractEmbeddedH2OConfig c) {
+    embeddedH2OConfig = c;
+  }
+
+  /**
+   * Returns an instance of {@link AbstractEmbeddedH2OConfig}. The origin of the embedded config might be either
+   * from directly setting the embeddedH2OConfig field via setEmbeddedH2OConfig setter, or dynamically provided via
+   * service loader. Directly set {@link AbstractEmbeddedH2OConfig} is always prioritized. ServiceLoader lookup is only
+   * performed if no config is previously set.
+   * <p>
+   * Result of first ServiceLoader lookup is also considered final - once a service is found, dynamic lookup is not
+   * performed any further.
+   *
+   * @return An instance of {@link AbstractEmbeddedH2OConfig}, if set or dynamically provided. Otherwise null
+   * @author Michal Kurka
+   */
+  public static AbstractEmbeddedH2OConfig getEmbeddedH2OConfig() {
+    if (embeddedH2OConfig != null) {
+      return embeddedH2OConfig;
+    }
+
+    final Optional<EmbeddedConfigProvider> optionalEmbeddedConfigProvider = discoverEmbeddedConfigProvider();
+
+    if (optionalEmbeddedConfigProvider.isPresent()) {
```

Comment:
would be nicer with map().orElse()

Reply:
Changed. Nicely reveals how probably even this getter could output Optional - as it's often null. 

@michalkurka Wanted to incorporate some changes himself as well.